File: Golang/test/valid_parenthesis.go
package test

func IsValid(s string) bool {
	match := map[rune]rune{ // Map untuk cocokin kurung tutup -> kurung buka
		')': '(',
		']': '[',
		'}': '{',
	}

	stack := []rune{} // Slice di Go berfungsi sebagai Stack (push/pop)

	
	for _, char := range s { // Loop setiap karakter (rune)
		if openBracket, ok := match[char]; ok { // Cek: Apakah char adalah Kurung TUTUP?
			if len(stack) == 0 || stack[len(stack)-1] != openBracket { // 1.) Cek Stack kosong? (Ada kurung tutup tanpa kurung buka) 2.) Cek elemen teratas Stack cocok gak sama pasangannya?
				return false
			}
	
			stack = stack[:len(stack)-1] // Kalau cocok, Pop/Keluarkan elemen teratas dari Stack
		} else {
			stack = append(stack, char) // Kalau char adalah Kurung BUKA, Push/Masukkan ke Stack
		}
	}

	return len(stack) == 0 // Valid jika dan hanya jika setelah selesai loop, Stack KOSONG
}

File: Javascript/day_2_stack_and_greedy/validParenthesis.js
var isValid = function(s) {
    const stack = []; // Stack untuk menyimpan kurung buka

    for (let i = 0; i < s.length; i++) { // Loop setiap karakter
        const char = s[i];

        if (char === '(' || char === '{' || char === '[') { // Jika ketemu kurung buka, simpan ke stack
            stack.push(char);
        } else { // Jika ketemu kurung tutup

            if (stack.length === 0) {   // Jika stack kosong tapi ada kurung tutup
                return false; // artinya tidak ada pasangan -> INVALID
            }

            const top = stack.pop();  // Ambil elemen terakhir dari stack (LIFO)

            // Cek apakah jenis kurung cocok
            if (char === ')' && top !== '(') return false;
            if (char === '}' && top !== '{') return false;
            if (char === ']' && top !== '[') return false;
        }
    }

    // Jika masih ada kurung buka yang belum ditutup
    return stack.length === 0;
};
